calculator.rb

```diff
@@ -0,0 +1,170 @@
+#   1. a math operation
+#   2. a number for the operation
+#   3. a second number for the operation
+
+# evaluate the arithmetic of the input.
+# output evaluation to user.
+
+# should support:
+#   1. addition ("add" or "+")
+#   2. subtraction ("subtract" or "-")
+#   3. multiplication ("multiply" or "*")
+#   4. division ("divide" or "/")
+# (you may add other names/symbols for these)
+
+# make sure program handles:
+#   all supported arithmetic
+#   dividing by zero
+#   erroneous input (exit the program when invalid input is entered)
+
+# optional enhancements:
+#   keep asking for valid input until it is received. [done]
+#   print out the formula in addition to the result. [done]
+#   add support for computing exponents. [done]
+#   add support for the modulo operator. [done]
+#   make your program know when it needs to return an integer vs a float. [done]
+#   add support for parentheticals. []
+#   add support for handing uppercase. [done]
+
+
+## Variables
+
+welcome_to_program = "Welcome to the calculator program!\n"
+welcome_to_program << "You will be asked to input a math operator and two numbers to be operated upon.\n"
+
+# possible arithmetic commands/operators.
+commands = {
+  :add => ["add", "+"],
+  :subtract => ["subtract", "-"],
+  :multiply => ["multiply", "*"],
+  :divide => ["divide", "/"],
+  :modulo => ["modulo", "%"],
+  :exponent => ["exponent", "**"]
+}
+
+# holds the desired arithmetic ingredients to be gathered from the user.
+desired_inputs = {
+  :operator => "mathematical operator",
+  :first_num => "number",
+  :second_num => "second number"
+}
+
+# calculation ingredients gathered from user.
+user_ingredients = {}
+
+
+## Methods
+
+# gather all the calculation ingredients from the user.
+def ask_for_ingredients(ingredients, user_ingredients, commands)
+  
+  # for each of the ingredients, ask the user for input.
+  ingredients.each_pair do |type_name, input_type|
+    
+    while user_ingredients[type_name] == nil
+      user_input = ask_for_input(input_type)
+      
+      if type_name == :second_num && commands[:divide].include?(user_ingredients[:operator]) && is_zero(user_input)
+        puts "You cannot divide by zero."
+      elsif (type_name == :operator && !is_command(user_input, commands)) ||
+        (type_name == :first_num && !is_numeric(user_input)) ||
+        (type_name == :second_num && !is_numeric(user_input)) 
+        puts "You did not enter an #{input_type}"
+      else
+        user_ingredients[type_name] = user_input
+      end
+    end
+    
+  end
+  
+end
+
+# asks the user to input a single piece of data.
+def ask_for_input(input_type)
+  puts prompt = "Please enter a #{input_type}:  "
+  user_input = gets.chomp.downcase
+  return user_input
+end
+
+# checks if the user's input is a number.
+def is_numeric(str)
+  return !!(str =~ /\A[-]*\d+\Z/ || str =~ /\A[-]*\d+\.\d+\Z/)
+end
+
+# checks if the user's input is a zero.
+def is_zero(str)
+  return !!(str =~ /\A[-]*[0]\Z/)
```

Comment:
I'm not sure that this complex of a regular expression is necessary here. Is there another way you could check for zero?